Added celery worker

rusko124 opened the pull request:
Discussion here #876
This is just a sketch for how Screenly can work with Celery.
At the moment, no tasks and support for Docker and Resin.io

vpetersson commented:
@rusko124 good start, but we're missing a few things #734 in the skelleton design.

rusko124 commented:
I will combine this question #734, as it depends on Celery.
In this large commit, I used notes from https://github.com/Screenly/screenly-ose/issues/734#issuecomment-441737456
I'll leave comments about those or other points in the code.

The problem is updating when we enable password prompt for sudo. This is solved by running the update with sudo, but without it we will get errors. The update script was also corrected.

I understand that we still need changes, but now we can update through the web interface(just only for master branch).

vpetersson commented on ansible/roles/screenly/tasks/main.yml:
```diff
@@ -70,6 +70,14 @@
     job: "/usr/local/bin/screenly_utils.sh cleanup"
     user: pi
 
+- name: Copy upgrade_screenly.sh
```
This is an attack vector. A user could simply add any arbitrary comand into this file and now they can execute anything as root.

rusko124 replied:
Updated

vpetersson commented on ansible/roles/system/tasks/main.yml:
```diff
@@ -241,3 +240,18 @@
     path: /var/swap
     state: absent
+
+- name: Remove pi-nopasswd file
```
👍

vpetersson commented on ansible/roles/system/files/010_pi-nopasswd:
```diff
@@ -0,0 +1 @@
+pi ALL=(ALL) NOPASSWD: ALL
```
This is wrong. We should require password. We then need to whitelist selected few commands.

vpetersson replied:
Fair enough, but we still need another one then for the locked down state.

rusko124 replied:
The task above just deletes this file (010_pi-nopasswd file for disabling sudo password in raspbian by default)

rusko124 commented:
So that I added above.
- I changed celery broker from rabbitmq to redis
- Now the update script that is used through the web interface is taken directly from the repository
- Added docker-compose for celery health in Docker environment.
- Added support for celery in resin.io
- Added authentication to run system commands through the web interface. This will provide greater security. This is also needed for this #121. You cannot run these commands without enabling this option.

vpetersson commented on bin/upgrade_screenly.sh:
```diff
@@ -0,0 +1,90 @@
+#    export DOCKER_TAG="production"
+#    BRANCH="production"
+else
+  exit 1
```
Maybe add some logging here?

vpetersson commented:
Looking good, @rusko124. What's left in the initial scope?

rusko124 commented:
@vpetersson 
Um, I think we just need to work on it:
- run_upgrade.sh for manual start
- upgrade_screenly.sh to update via web interface

Will we use run_upgrade.sh script through sudo or something else better?

vpetersson commented on Dockerfile.worker:
```diff
@@ -0,0 +1,17 @@
+FROM python:2.7-stretch
+MAINTAINER Viktor Petersson <vpetersson@screenly.io>
```
This is a new file. You'd be the maintainer ;)

rusko124 replied:
Got it :)

rusko124 commented:
List of changes in the last commit:
- Instead of 2 files updates now is one
- Manual run of the update for the user remained unchanged
- Setting a required sudo password is now directly in the script itself, since it is not possible to do this in the ansible.
- Some edits

rusko124 commented:
@vpetersson Could you look it up when you'll have time? And if everything is as you would like to see it then I will work on the quality of the code.

vpetersson commented on server.py:
```diff
@@ -40,20 +42,50 @@
 
 HOME = getenv('HOME', '/home/pi')
 DISABLE_MANAGE_NETWORK = '.screenly/disable_manage_network'
+CELERY_RESULT_BACKEND = getenv('CELERY_RESULT_BACKEND', 'redis://localhost:6379/0')
```
Is there a scenario where `CELERY_RESULT_BACKEND` and `CELERY_BROKER_URL` are different? If not, I suggest we consolidate them to `CELERY_BROKER`

rusko124 replied:
Hmm I think that variables likely will not different, but the result and broker have a different purposes. So I think we don't need to merger them into one variable, it can be confusing.

rusko124 commented:
@vpetersson 
Could you look at this commit https://github.com/Screenly/screenly-ose/pull/987/commits/9c6888da425a5b57d8e7cccde62de8f6267944f2 ?
There I have combined a few issues:
- https://github.com/Screenly/screenly-ose/issues/876
- https://github.com/Screenly/screenly-ose/pull/897#issuecomment-419362352
- Added stop and pause commands for the viewer
- Removed screenly_utils.sh 
- Added support for periodic tasks to celery

vpetersson commented:
@rusko124 Let's createa  new branch so that we can test it easier. It's a pretty large change so we should thread lightly.
